modules/message-lane/src/lib.rs

```diff
@@ -0,0 +1,283 @@
+			let _ = ensure_signed(origin)?;
+			let mut lane = outbound_lane::<T, I>(lane_id);
+			let nonce = lane.send_message(payload);
+			lane.prune_messages(T::MaxMessagesToPruneAtOnce::get());
+
+			Self::deposit_event(Event::MessageAccepted(lane_id, nonce));
+		}
+	}
+}
+
+impl<T: Trait<I>, I: Instance> Module<T, I> {
+	// =========================================================================================
+	// === Exposed mutables ====================================================================
+	// =========================================================================================
+
+	/// Receive new TRUSTED lane messages.
+	///
+	/// Trusted here means that the function itself doesn't check whether message has actually
+	/// been sent through the other end of the channel. We only check that we are receiving
+	/// and processing messages in order here.
+	///
+	/// Messages vector is required to be sorted by nonce within each lane. Otherise messages
+	/// will be rejected.
+	pub fn receive_messages(messages: Vec<Message<T::Payload>>) -> MessageNonce {
+		let mut correct_messages = 0;
+		let mut processor = T::OnMessageReceived::default();
+		for message in messages {
+			let mut lane = inbound_lane::<T, I>(message.key.lane_id);
+			if lane.receive_message(message.key.nonce, message.payload, &mut processor) {
+				correct_messages += 1;
+			}
+		}
+
+		correct_messages
+	}
+
+	/// Process stored lane messages.
+	///
+	/// Stops processing either when all messages are processed, or when processor returns
+	/// MessageResult::NotProcessed.
+	pub fn process_lane_messages(lane_id: &LaneId, processor: &mut impl OnMessageReceived<T::Payload>) {
+		inbound_lane::<T, I>(*lane_id).process_messages(processor);
+	}
+
+	/// Receive TRUSTED proof of message receival.
+	///
+	/// Trusted here means that the function itself doesn't check whether the bridged chain has
+	/// actually received these messages.
+	///
+	/// The caller may break the channel by providing `latest_received_nonce` that is larger
+	/// than actual one. Not-yet-sent messages may be pruned in this case.
+	pub fn confirm_receival(lane_id: &LaneId, latest_received_nonce: MessageNonce) {
+		let mut lane = outbound_lane::<T, I>(*lane_id);
+		let received_range = lane.confirm_receival(latest_received_nonce);
+		lane.prune_messages(T::MaxMessagesToPruneAtOnce::get());
+
+		if let Some(received_range) = received_range {
+			Self::deposit_event(Event::MessagesDelivered(*lane_id, received_range.0, received_range.1));
+		}
+	}
+
+	/// Receive TRUSTED proof of message processing.
+	///
+	/// Trusted here means that the function itself doesn't check whether the bridged chain has
+	/// actually processed these messages.
+	pub fn confirm_processing(lane_id: &LaneId, latest_processed_nonce: MessageNonce) {
+		let mut lane = outbound_lane::<T, I>(*lane_id);
+		let processed_range = lane.confirm_processing(latest_processed_nonce);
+		lane.prune_messages(T::MaxMessagesToPruneAtOnce::get());
```

Comment:
Why do we prune both here and when sending messages? Wouldn't one of these be enough?

Reply:
Actually I wonder if we should rather amortize pruning cost and do it during `on_initialize`. I.e. the relayer should pay a fee (weight) when the message is queued that should include the pruning cost as well. Then the pruning can happen in `on_initialize`.

Otherwise pruning randomly will make `weight` calculations a nightmare and will make the user experience quite unpredictable.

Reply:
I was in doubts about where it is more appropriate to do that - see last point in PR description.The final idea is that "if you want to use the lane, then you should pay for its maintenance". So it is called from every place where (outbound) lane state is updated - i.e. whenever message is sent, or we confirm receiving/processing.

I'm not sure I understand your point about nightmare. `prune_messages()` has upper bound - it is: `(1 + T::MaxMessagesToPruneAtOnce::get()`)  writes. Afaiu, we just include that upper bound into default call weight and then return actual weight if less than `MaxMessagesToPruneAtOnce` headers has been pruned.

Re pruning from `on_initialize`/`on_finalize` - the main disadvantage is that we either need parallel structure (map?) to track lanes that actually need pruning, or we need to enumerate all lanes in the storage. Both options may end up with potentially large amount of reads (+writes sometimes) => slow down block production/import time, probably even if nothing needs to be pruned.

In any case - imo "amortized pruning" is exactly what's currently happening - i.e. we amortize cost of pruning between all lane users, not even all blockchain users.

Does it makes sense, or you're still convinced that pruning from on_initialize/on_finalize is better?

Reply:
> I'm not sure I understand your point about nightmare.  `prune_messages()` has upper bound

I've said "nightmare", because I meant exact calculations. Over-estimating and using an upper bound here will make a sub-optimal user experience and block utilisation. Assuming that queueing a message is ~2/3 state writes, and `MaxMessagesToPruneAtOnce` is going to be set to dozens, we may end up with quite high over estimation, and extra code to calculate refunds. I'd rather focus on lowering the cost (and it's variance) for relayers and users.

> Re pruning from on_initialize/on_finalize - the main disadvantage is that we either need parallel structure (map?) to track lanes that actually need pruning, or we need to enumerate all lanes in the storage. 

That's a good point and I really don't have the right answer. The best strategy here actually depends on whether we will have many in-frequently used lanes or rather a small amount of very frequently used ones. My main critique towards `MaxMessagesToPruneAtOnce` and pruning-when-used is that it makes actual costs of some operations a bit unpredicatble (although bonded), so it creates some unnecessary fee dynamics (like back-running a message to prevent paying the pruning costs), which might be significant if the chain fees are high.


That said, I think I'm convinced that it's better to prune-when-used, but I think we should only do it on `send_message`. Here is why. Imagine that we set `MaxMessagesToPruneAtOnce` to some large value. If the relayers are slow (or the other chain is slow), senders will keep adding more and more messages without really pruning anything. Then whenever this situation get's resolved, the relayers will be required to pay a high cost of pruning, when messages are received - without proper incentivisation they may simply stop doing this. There isn't much they loose really.
However senders, should be prevented from sending too many messages that are not properly processed (rate limitting), so the above situation should never really happen. Therefore I'd rather have relayers confirm delivery cheaply and then have senders prune old messages & being rate-limited at the same time. Since, rate-limitting should (or may) cause an (exponentialy) increasing cost of sending a message (which basically creates some upper limit), we may also use up this fee to prune more and more entries, hence the number of messages that gets pruned during send would go up along the undelivered queue size.
At worst we will end up with a bunch of unpruned messages in infrequently used lanes. To cater for this we might consider making the minimal number of messages to prune depenendent on the usage frequency as well. I.e. if the lane is used frequently, `send_message` prunes like at most 1 or 2 entries, if there was no messages for couple of blocks we will prune significantly more old messages on the next `send_message`.

So again, TL;DR I agree it's better to leave as-is for now (maybe without pruning on confirmation though), making `MaxMessagesToPruneAtOnce` dynamic in the future will prevent more sophisticated problems, but pruning in `on_initialize` doesn't really seem to solve any issues either.

Reply:
Ok, removed pruning from confirmation calls